lib/matplotlib/axis.py

```diff
@@ -1626,14 +1631,24 @@ def set_ticklabels(self, ticklabels, *, minor=False, **kwargs):
         """
         ticklabels = [t.get_text() if hasattr(t, 'get_text') else t
                       for t in ticklabels]
+        locator = (self.get_minor_locator() if minor
+                   else self.get_major_locator())
+        if isinstance(locator, mticker.FixedLocator):
+            tickd = {loc: lab for loc, lab in zip(locator.locs, ticklabels)}
+            formatter = mticker.FuncFormatter(self._format_with_dict, tickd)
+        else:
+            formatter = mticker.FixedFormatter(ticklabels)
         if minor:
-            self.set_minor_formatter(mticker.FixedFormatter(ticklabels))
+            self.set_minor_formatter(formatter)
             ticks = self.get_minor_ticks()
         else:
-            self.set_major_formatter(mticker.FixedFormatter(ticklabels))
+            self.set_major_formatter(formatter)
             ticks = self.get_major_ticks()
+
+        self._update_ticks()
```

Comment:
do you really need to call update_ticks() here?  and if you do, shouldn't this go *before* get_major_ticks()/get_minor_ticks()?  (otherwise you'd be possibly invalidating `ticks` even before iterating on them...)